cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go

```diff
@@ -0,0 +1,226 @@
+	"fmt"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	"k8s.io/kubernetes/cmd/kubeadm/app/cmd/options"
+	"k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow"
+	cmdutil "k8s.io/kubernetes/cmd/kubeadm/app/cmd/util"
+	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	certsphase "k8s.io/kubernetes/cmd/kubeadm/app/phases/certs"
+	"k8s.io/kubernetes/cmd/kubeadm/app/phases/controlplane"
+	etcdphase "k8s.io/kubernetes/cmd/kubeadm/app/phases/etcd"
+	kubeconfigphase "k8s.io/kubernetes/cmd/kubeadm/app/phases/kubeconfig"
+	kubeconfigutil "k8s.io/kubernetes/cmd/kubeadm/app/util/kubeconfig"
+)
+
+type controlPlanePrepareData interface {
+	Cfg() *kubeadmapi.JoinConfiguration
+	InitCfg() (*kubeadmapi.InitConfiguration, error)
+}
+
+func getControlPlanePreparePhaseFlags() []string {
+	return []string{
+		options.APIServerAdvertiseAddress,
+		options.APIServerBindPort,
+		options.CfgPath,
+		options.ControlPlane,
+		options.NodeName,
+		options.TokenDiscovery,
+		options.TokenDiscoveryCAHash,
+		options.TokenDiscoverySkipCAHash,
+	}
+}
+
+// NewControlPlanePreparePhase creates a kubeadm workflow phase that implements the preparation of the node to serve a control plane
+func NewControlPlanePreparePhase() workflow.Phase {
+	return workflow.Phase{
+		Name:  "control-plane-prepare",
+		Short: "Prepares the machine for serving a control plane.",
+		Long:  cmdutil.MacroCommandLongDescription,
+		Phases: []workflow.Phase{
+			newCertsSubphase(),
+			newKubeconfigSubphase(),
+			newManifestsSubphases(),
+			newCheckLocalEtcdStatusSubphase(),
+		},
+	}
+}
+
+func newCertsSubphase() workflow.Phase {
+	return workflow.Phase{
+		Name:         "certs",
+		Short:        "Generates the certificates for the new control plane components",
+		Run:          runCertsPhaseLocal,
+		InheritFlags: getControlPlanePreparePhaseFlags(),
+	}
+}
+
+func newKubeconfigSubphase() workflow.Phase {
+	return workflow.Phase{
+		Name:         "kubeconfig",
+		Short:        "Generates the kubeconfig for the new control plane components",
+		Run:          runKubeconfigPhaseLocal,
+		InheritFlags: getControlPlanePreparePhaseFlags(),
+	}
+}
+
+func newManifestsSubphases() workflow.Phase {
+	return workflow.Phase{
+		Name:  "manifests",
+		Short: "Generates the manifests for the new control plane components",
+		Phases: []workflow.Phase{
+			{
+				Name:           "all",
+				Short:          "Generates all static Pod manifest files",
+				InheritFlags:   getControlPlanePreparePhaseFlags(),
+				RunAllSiblings: true,
+			},
+			newControlPlanePrepareSubPhase(kubeadmconstants.KubeAPIServer),
+			newControlPlanePrepareSubPhase(kubeadmconstants.KubeControllerManager),
+			newControlPlanePrepareSubPhase(kubeadmconstants.KubeScheduler),
+		},
+		InheritFlags: getControlPlanePreparePhaseFlags(),
+	}
+}
+
+func newCheckLocalEtcdStatusSubphase() workflow.Phase {
+	return workflow.Phase{
+		Name:         "check-local-etcd-status",
```

Comment:
this should have been part of @MalloZup 's phase (preflight)
looks like we've missed it.

https://github.com/kubernetes/kubeadm/issues/1204
> etcd check (if initConfiguration.Etcd.External == nil from PrepareForHostingControlPlane)

@ereslibre could you please move it as part of this PR to preflight.go?

Reply:
might be a problem, yes...
please check, but if that is the case we should leave it here for now.

Reply:
I'm not sure this can easily moved on preflight due to the dependency defined above, let's chat about this

Reply:
So, for checking the etcd health we are using the etcd healthcheck certificate and key that are not copied among masters, but generated automatically on every master. So I'd say we should keep this subphase after the PKI generation in this phase.